File: node_js_ds_and_algos_implementations/trees/level_order_traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */

/* 
A few quirks with the JS implementation vs the Python Implementation
1) Using an array as a queue (since there's no native queue implementation in JS),
and doing shift(), which pops the leftmost element of the array
2) Make sure to store the original length of the queue before entering the for loop,
and adding items to the queue.
*/
var levelOrder = function(root) {
    if (!root){
        return []
    }
    let q = []
    let levels = []
    q.push(root)
    while (q.length){
        let level = []
        let qLength = q.length
        for (let i = 0; i < qLength; ++i){
            let node = q.shift()
            if (node){          
                if (node.left){
                    q.push(node.left)
                }
                if (node.right){
                    q.push(node.right)
                }
                level.push(node.val)
            }
        }
        levels.push(level)
    }
    return levels
};

File: neetcode java implementations/LevelOrderTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        if (root == null){
            List<List<Integer>> empty = new ArrayList<>();
            return empty;
        }
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        List<List<Integer>> res = new ArrayList<>();
        while (!q.isEmpty()){
            int N = q.size();
            List<Integer> level = new ArrayList<Integer>();
            for (int i = 0; i < N; ++i){
                TreeNode node = q.poll();
                if (node != null){
                    if (node.left != null){
                        q.add(node.left);
                    }
                    if (node.right != null){
                        q.add(node.right);
                    }
                    level.add(node.val);
                }
            }
            res.add(level);
        }
        return res;
    }
}
